lib/streamlit/elements/multiselect.py

```diff
@@ -264,6 +277,20 @@ def _multiselect(
             serializer=serde.serialize,
             ctx=ctx,
         )
+
+        if max_selections is not None:
+            if get_default_count(widget_state.value) > max_selections:
+                raise StreamlitAPIException(
+                    f"""
+Multiselect has {get_default_count(widget_state.value)} options selected but `max_selections`
+is set to {max_selections}. This happened because you either gave too many options to `default`
+or you manipulated the widget's state through `st.session_state`. Note that
+the latter can happen before the line indicated in the traceback.
+Please select at most {max_selections} options.
+"""
+                )
+            else:
+                multiselect_proto.max_selections = max_selections
```

Comment:
Hm, so I think where we choose to set `multiselect_proto.max_selections = max_selections` ends up having some weird product implications that are worth considering.

If we set the value here, then changing it won't affect a widget's ID (as widget IDs are generated in the call to `register_widget`), and this means we'll have yet another way of ending up with a multiselect with more elements in it than `max_selections`:

1. Create a multiselect with a `max_selections` value
2. Pick a bunch of multiselect options
3. Decrease the `max_selections` value to be less than the number of currently selected options


If we set it above (before the `if help is not None` statement), then changing `max_selections` will reset the currently selected multiselect options (because we're incorporating `max_selections` into the multiselect ID, so we'll technically have a new multiselect from the perspective of the frontend), which will probably be in what the user wants in some cases.

With either option, we end up with some edge cases in some situations that I could see a user running into, and I'm not exactly sure what the best product behavior would be here. I think(?) the way things currently work is more reasonable since running into this error seems better than not being able to dynamically change `max_selections` without all of the current options being reset, but I'm not really sure about this.

(CC @jrieke for help on the product side)

Reply:
> In most cases we want widgets to reset when their parameters change

@vdonato @willhuang1997 Do we do this for all widget parameters right now? If no, which exceptions are there?

I'm actually leaning towards resetting the widget if `max_selections` changes. I can't imagine many use cases where it makes sense to dynamically change `max_selections`. If we reset everywhere else, then let's keep it consistent. (But maybe we do have exceptions for similar cases, so let's see which ones they are and try to be consistent with that!)

Reply:
Ok just played around with some parameters and we do also reset the widget e.g. for the `format_func` parameter, which just changes how different options are displayed. So if we reset the widget for that parameter, I think we should definitely reset for `max_selections`. 

Suggestion: one of you spends 10 minutes going through the code base looking for similar cases. If we don't find any, we reset the widget if `max_selections` changes.

Reply:
I thought about this a bit over the weekend, and I'd also lean toward resetting the widgets even though I'm sure that eventually, some people will run into the edge case and be confused. This is the behavior for essentially every other widget arg (aside from the one exception that I mention below), so unless we have good reason to believe that dynamically changing max selections and expecting the widget to retain its values is a relatively common thing to do with this argument, I don't think it's worth making another exception.

Resetting the widget is what happens the vast majority of the time, and the one exception that exists for all widgets (as far as I know) is the `disabled` parameter, where resetting the widget when that is toggled actually makes the parameter fairly useless.

Reply:
SG, then let's reset here!